### PR TITLE
[DATA] DataLoader -- a universal interface for wrapping datasets from other machine learning frameworks

### PR DESCRIPTION
**Pull request by electriclilies**
In this PR, I introduce the DataLoader class. The DataLoader is an abstract class intended to wrap datasets from other machine learning frameworks so that they can be used interchangeably within TVM for any data-aware tasks. 

I also provide three implementations of the DataLoader class: TFDataLoader NumpyDataLoader, and RandomDataLoader. 

The TFDataLoader wraps Tensorflow datasets. 
The NumpyDataLoader wraps numpy arrays of data in NCHW form (where N is the total number of datapoints). Keras datasets provide data in this form-- the Numpy DataLoader is intended for use with Keras datasets, but could also be used with any other dataset stored in a similar fashion. 
The RandomDataLoader takes in a list of shapes and produces random outputs that correspond to those shapes. This class is useful for testing code, especially if you are not at a point where you want to go to the effort of downloading a real dataset.

The DataLoader class was originally designed for data-aware quantization. I think it could also be useful for training, and for making accuracy testing scripts more robust and general.

@mbrookhart @jwfromm @altanh Please take a look and let me know what you think!

### CODE REVIEW

**Comment by jroesch** [comment]
@altanh does this work for PT dataloading?

**Comment by tqchen** [comment]
Thanks @electriclilies give this is an intro of a new feature, it would be great to send an RFC. 

There are quite a lot of previous practices (e.g. PyTorch, MXNet and TF's dataset API). It would be great to compare from these and make least amount of deviations so users can have a consistent usage experience

**Comment by electriclilies** [comment]
@tqchen I'll get an RFC up soon, and @jroesch I can double check that this would work with PT data loaders.

**Comment by electriclilies** [comment]
@tqchen @jroesch @anijain2305 I put up an RFC, please take a look: 
https://discuss.tvm.apache.org/t/dataloader-an-api-to-wrap-datasets-from-other-machine-learning-frameworks/9498
